pkg/supportbundle/supportbundle.go

```diff
@@ -0,0 +1,310 @@
+package supportbundle
+
+import (
+	"archive/zip"
+	"bytes"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+	"runtime"
+	"runtime/pprof"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/grafana/agent/pkg/build"
+	"github.com/grafana/agent/pkg/config"
+	"github.com/grafana/agent/pkg/traces"
+	zaplogfmt "github.com/jsternberg/zap-logfmt"
+	"github.com/mackerelio/go-osstat/uptime"
+	"github.com/sirupsen/logrus"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+	"gopkg.in/yaml.v3"
+)
+
+// Bundle collects all the data that is exposed as a support bundle.
+type Bundle struct {
+	meta                  []byte
+	config                []byte
+	agentMetrics          []byte
+	agentMetricsInstances []byte
+	agentMetricsTargets   []byte
+	agentLogsInstances    []byte
+	heapBuf               *bytes.Buffer
+	goroutineBuf          *bytes.Buffer
+	blockBuf              *bytes.Buffer
+	mutexBuf              *bytes.Buffer
+	cpuBuf                *bytes.Buffer
+}
+
+// Metadata contains general runtime information about the current Agent.
+type Metadata struct {
+	BuildVersion    string   `yaml:"build_version"`
+	OS              string   `yaml:"os"`
+	Architecture    string   `yaml:"architecture"`
+	Uptime          float64  `yaml:"uptime"`
+	EnabledFeatures []string `yaml:"enabled_features"`
+}
+
+// Export gathers the information required for the support bundle.
+func Export(enabledFeatures []string, cfg config.Config, srvAddress string, duration float64) (*Bundle, error) {
+	// Gather runtime metadata.
+	ut, err := uptime.Get()
+	if err != nil {
+		return nil, err
+	}
+	m := Metadata{
+		BuildVersion:    build.Version,
+		OS:              runtime.GOOS,
+		Architecture:    runtime.GOARCH,
+		Uptime:          ut.Seconds(),
+		EnabledFeatures: enabledFeatures,
+	}
+	meta, err := yaml.Marshal(m)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal support bundle metadata: %s", err)
+	}
+	// Gather current configuration.
+	config, err := yaml.Marshal(cfg)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal config: %s", err)
+	}
+
+	// Gather Agent's own metrics.
+	resp, err := http.DefaultClient.Get("http://" + srvAddress + "/metrics")
+	if err != nil {
+		return nil, fmt.Errorf("failed to get internal Agent metrics: %s", err)
+	}
+	agentMetrics, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read internal Agent metrics: %s", err)
+	}
+
```

Comment:
What about metrics from integrations? Or too much data?

Reply:
I initially left out the integrations as they required different handling.

With the current approach of using the middleware-y logger, we get (at least some) of the integrations' logs for free, so I'll check out how to get their metrics, too.